File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* root, vector<int> &v){
        if(!root) return;
        inorder(root->left, v);
        v.push_back(root->val);
        inorder(root->right, v);
    }
    bool isValidBST(TreeNode* root) {
        if(!root) return true;
        vector<int> v1,v2;
        inorder(root,v1);
        v2.assign(v1.begin(),v1.end());
        sort(v2.begin(), v2.end());
        if(v1==v2){
            for(int i=1;i<v2.size();i++){
                if(v2[i-1]==v2[i]) return false;
            }
            return true;
        }
        return false;
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean solve(TreeNode root, long mx, long mn) {
        if (root == null) return true;
            
        if((root.val>=mx) || (root.val<=mn)) return false;
        return solve(root.left, root.val, mn) && solve(root.right, mx, root.val);
    }

    public boolean isValidBST(TreeNode root) {
        
        return solve(root, Long.MAX_VALUE, Long.MIN_VALUE);
    }
}